ccl/sqlproxyccl/acl: TestAllowlistLogic failed

ccl/sqlproxyccl/acl.TestAllowlistLogic [failed](https://mesolite.cluster.engflow.com/invocations/default/e31feced-9334-4320-8e73-d3b043be0c21?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3NxbHByb3h5Y2NsL2FjbDphY2xfdGVzdA==) on release-25.3 @ [d596aa63eb1d22e88a5553ba6b333e8ad4f61ea5](https://github.com/cockroachdb/cockroach/commits/d596aa63eb1d22e88a5553ba6b333e8ad4f61ea5):


```
=== RUN   TestAllowlistLogic
    file_test.go:502: 
        	Error Trace:	pkg/ccl/sqlproxyccl/acl/file_test.go:502
        	Error:      	An error is expected but got nil.
        	Test:       	TestAllowlistLogic
E251118 09:07:20.218255 16618 ccl/sqlproxyccl/acl/file.go:106  [-] 12  WatchList daemon stopped: context canceled
--- FAIL: TestAllowlistLogic (0.61s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAllowlistLogic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57038

--- a/pkg/ccl/sqlproxyccl/acl/file_test.go
+++ b/pkg/ccl/sqlproxyccl/acl/file_test.go
@@ -493,15 +493,31 @@ allowlist:
 	for _, tc := range testCases {
 		require.NoError(t, os.WriteFile(filename, []byte(tc.input), 0777))
 
-		controller := <-channel
-		for _, ioPairs := range tc.specs {
-			err := controller.CheckConnection(ctx, ioPairs.connection)
-			if ioPairs.outcome == "" {
-				require.Nil(t, err)
-			} else {
-				require.EqualError(t, err, ioPairs.outcome)
+		// Keep polling until we get the controller we expect. It's possible
+		// that occasionally things run too slowly, and a stale controller
+		// without the most recent data gets pushed because the polling
+		// interval passes. However, we do know that eventually we expect to
+		// read our desired controller.
+		testutils.SucceedsSoon(t, func() error {
+			select {
+			case controller := <-channel:
+				for _, ioPairs := range tc.specs {
+					err := controller.CheckConnection(ctx, ioPairs.connection)
+					if ioPairs.outcome == "" {
+						if err != nil {
+							return errors.Newf("expected no error, but got %v", err) // nolint:errwrap
+						}
+					} else {
+						if err == nil || err.Error() != ioPairs.outcome {
+							return errors.Newf("expected %v, but got %v", ioPairs.outcome, err) // nolint:errwrap
+						}
+					}
+				}
+				return nil
+			default:
+				return errors.New("no data")
 			}
-		}
+		})
 	}
 }
 

